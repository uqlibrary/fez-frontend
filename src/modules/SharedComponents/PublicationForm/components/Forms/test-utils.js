import { useForm } from 'react-hook-form';
import { render, WithReduxStore } from 'test-utils';
import React from 'react';
import { formValues } from '../PublicationForm';

export const setupWrapper = (Form, props) => {
    const Wrapper = () => {
        const {
            control,
            formState: { isSubmitting },
        } = useForm();

        // eslint-disable-next-line react/prop-types
        return <Form {...{ control, isSubmitting, ...props, formValues: formValues(props?.formValues || {}) }} />;
    };

    return render(
        <WithReduxStore>
            <Wrapper />
        </WithReduxStore>,
    );
};
