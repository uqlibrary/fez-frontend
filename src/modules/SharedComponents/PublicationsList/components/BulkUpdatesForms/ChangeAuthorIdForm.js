import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import { getFormSyncErrors, Field, reduxForm, SubmissionError, formValueSelector, change } from 'redux-form/immutable';

import { Alert } from 'modules/SharedComponents/Toolbox/Alert';
import { UqIdField } from 'modules/SharedComponents/LookupFields';
import { NewGenericSelectField } from 'modules/SharedComponents/GenericSelectField';
import { TextField as GenericTextField } from 'modules/SharedComponents/Toolbox/TextField';
import Grid from '@mui/material/Grid';
import Button from '@mui/material/Button';

import { locale } from 'locale';
import { validation } from 'config';
import { changeAuthorId } from 'actions';
import { SEARCH_AUTHOR_BY, SEARCH_BY_AUTHOR_NAME, SEARCH_BY_AUTHOR_ID } from 'config/bulkUpdates';

const FORM_NAME = 'ChangeAuthorIdForm';
const selector = formValueSelector(FORM_NAME);

const onSubmit = (values, dispatch, props) => {
    return dispatch(changeAuthorId(Object.values(props.recordsSelected), values.toJS())).catch(error => {
        throw new SubmissionError({ _error: error.message });
    });
};

export const ChangeAuthorIdForm = ({ error, handleSubmit, recordsSelected, submitting, submitSucceeded, onCancel }) => {
    const dispatch = useDispatch();
    const txt = locale.components.bulkUpdates.bulkUpdatesForms;
    const formErrors = useSelector(state => getFormSyncErrors(FORM_NAME)(state));
    const searchAuthorBy = useSelector(state => selector(state, 'search_author_by'));
    const searchAuthorByName = useSelector(state => selector(state, 'search_author.author'));

    const disableSubmit = !!formErrors && !(formErrors instanceof Immutable.Map) && Object.keys(formErrors).length > 0;
    const handleClear = field => () => dispatch(change(FORM_NAME, field, null));

    const authorNames = React.createRef(null);
    const [authorNameNoMatchCount, setAuthorNameNoMatchCount] = React.useState(null);
    authorNames.current = Object.values(recordsSelected).map(record =>
        record.fez_record_search_key_author.map(author => author.rek_author),
    );

    React.useEffect(() => {
        if (!!searchAuthorByName) {
            const count = authorNames.current.filter(author => !author.includes(searchAuthorByName)).length;
            setAuthorNameNoMatchCount(count);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [searchAuthorByName]);

    React.useEffect(() => {
        if (submitSucceeded) {
            setTimeout(onCancel, 2000);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [submitSucceeded]);

    return (
        <form data-testid="change-author-id-form" id="change-author-id-form">
            <Grid container spacing={2}>
                <Grid item xs={12}>
                    <Alert alertId="alert-info-change-author-id" {...txt.changeAuthorIdForm.alert} />
                </Grid>
                {searchAuthorBy === SEARCH_BY_AUTHOR_NAME && !!searchAuthorByName && (
                    <Grid item xs={12}>
                        <Alert
                            alertId="alert-warning-change-author-id"
                            {...txt.changeAuthorIdForm.warningAlert}
                            message={txt.changeAuthorIdForm.warningAlert.message
                                .replace('[authorNameNoMatchCount]', authorNameNoMatchCount)
                                .replace('[numberOfSelectedWorks]', Object.keys(recordsSelected).length)}
                        />
                    </Grid>
                )}
                <Grid item xs={12}>
                    <Field
                        component={NewGenericSelectField}
                        disabled={submitting || submitSucceeded}
                        genericSelectFieldId="search-author-by"
                        label={txt.changeAuthorIdForm.formLabels.searchBy}
                        itemsList={Object.values(SEARCH_AUTHOR_BY)}
                        name="search_author_by"
                        selectPrompt={txt.changeAuthorIdForm.selectPrompt}
                        required
                        validate={[validation.required]}
                    />
                </Grid>
                {searchAuthorBy === SEARCH_BY_AUTHOR_NAME && (
                    <Grid item xs={12}>
                        <Field
                            component={GenericTextField}
                            disabled={submitting || submitSucceeded}
                            textFieldId="search-by-rek-author"
                            fullWidth
                            label={txt.changeAuthorIdForm.formLabels.searchByAuthorName}
                            name={`search_author.${searchAuthorBy}`}
                            required
                            validate={[validation.required]}
                        />
                    </Grid>
                )}
                {searchAuthorBy === SEARCH_BY_AUTHOR_ID && (
                    <Grid item xs={12}>
                        <Field
                            component={UqIdField}
                            disabled={submitting || submitSucceeded}
                            floatingLabelText={txt.changeAuthorIdForm.formLabels.searchByAuthorId}
                            name={`search_author.${searchAuthorBy}`}
                            required
                            validate={[validation.required]}
                            uqIdFieldId="search-by-rek-author-id"
                            getOptionLabel={option => !!option && `${option.id} (${option.value})`}
                            normalize={value => value.aut_id}
                            onClear={handleClear(`search_author.${searchAuthorBy}`)}
                        />
                    </Grid>
                )}
                <Grid item xs={12}>
                    <Field
                        component={UqIdField}
                        disabled={submitting || submitSucceeded}
                        floatingLabelText={txt.changeAuthorIdForm.formLabels.authorId}
                        name="rek_author_id"
                        required
                        validate={[validation.required]}
                        uqIdFieldId="rek-author-id"
                        getOptionLabel={option => !!option && `${option.id} (${option.value})`}
                        normalize={value => value.aut_id}
                        onClear={handleClear('rek_author_id')}
                    />
                </Grid>
                <Grid item xs={12} sm={6}>
                    <Button
                        aria-label={txt.changeAuthorIdForm.formLabels.cancelButtonLabel}
                        children={txt.changeAuthorIdForm.formLabels.cancelButtonLabel}
                        data-analyticsid="change-author-id-cancel"
                        data-testid="change-author-id-cancel"
                        disabled={submitting}
                        fullWidth
                        id="change-author-id-cancel"
                        onClick={onCancel}
                        variant="contained"
                    />
                </Grid>
                <Grid item xs={12} sm={6}>
                    <Button
                        aria-label={txt.changeAuthorIdForm.formLabels.submitButtonLabel}
                        children={txt.changeAuthorIdForm.formLabels.submitButtonLabel}
                        color="primary"
                        data-analyticsid="change-author-id-submit"
                        data-testid="change-author-id-submit"
                        disabled={submitting || disableSubmit || submitSucceeded}
                        fullWidth
                        id="change-author-id-submit"
                        onClick={handleSubmit}
                        variant="contained"
                    />
                </Grid>
                <Grid item xs={12}>
                    {!!submitting && (
                        <Alert alertId="alert-info-change-author-id" {...txt.changeAuthorIdForm.submittingAlert} />
                    )}
                    {!!submitSucceeded && (
                        <Alert alertId="alert-done-change-author-id" {...txt.changeAuthorIdForm.successAlert} />
                    )}
                    {!!error && <Alert alertId="alert-error-change-author-id" {...txt.changeAuthorIdForm.errorAlert} />}
                </Grid>
            </Grid>
        </form>
    );
};

ChangeAuthorIdForm.propTypes = {
    error: PropTypes.string,
    handleSubmit: PropTypes.func,
    onCancel: PropTypes.func,
    recordsSelected: PropTypes.object,
    submitting: PropTypes.bool,
    submitSucceeded: PropTypes.bool,
};

const ChangeAuthorIdReduxForm = reduxForm({
    form: FORM_NAME,
    onSubmit,
})(ChangeAuthorIdForm);

export default React.memo(ChangeAuthorIdReduxForm);
