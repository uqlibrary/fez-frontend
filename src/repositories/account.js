import {api} from 'config';

/**
 * Fetches the user's account
 * @returns {Promise}
 */
export function getAccount() {
    return new Promise((resolve, reject) => {
        api.get('account').then(response => {
            if (response.data && response.data.hasOwnProperty('hasSession') && response.data.hasSession === true) {
                resolve(response.data);
            } else {
                reject({
                    response: {
                        status: 401,
                        data: 'Unauthorized user'
                    },
                });
            }
        }).catch(e => {
            console.dir(e);

            reject(e);

            // if (e.hasOwnProperty('response') && e.response !== null && typeof(e.response) !== 'undefined'
            //     && e.response.hasOwnProperty('status') && (e.response.status === 401 || e.response.status === 403)) {
            // } else {
            //     throw e;
            // }
        });
    });
}
