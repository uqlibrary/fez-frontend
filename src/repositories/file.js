import { generateCancelToken } from 'config';
import {
    MIME_TYPE_WHITELIST,
    FILE_ACCESS_CONDITION_CLOSED,
    FILE_ACCESS_CONDITION_OPEN,
    FILE_ACCESS_CONDITION_INHERIT,
    FILE_SECURITY_POLICY_PUBLIC,
} from 'modules/SharedComponents/Toolbox/FileUploader/config';
import * as fileUploadActions from 'modules/SharedComponents/Toolbox/FileUploader/actions';
import { FILE_UPLOAD_API } from './routes';
import { post, put } from './generic';
import * as Sentry from '@sentry/react';
import locale from 'locale/global';
const moment = require('moment');

export const getFileUploadMetadata = (file, collections) => {
    const securityInherited = !!file.access_condition_id && file.access_condition_id === FILE_ACCESS_CONDITION_INHERIT;
    let securityPolicy = file.security_policy;
    if (!!!securityPolicy) {
        // assume file_access is being set
        securityPolicy = file.access_condition_id;
        if (securityInherited) {
            const parentPolicy = collections.reduce(
                (policy, collection) =>
                    collection.rek_datastream_policy < policy ? collection.rek_datastream_policy : policy,
                FILE_ACCESS_CONDITION_OPEN,
            );
            securityPolicy = parentPolicy;
        } else if (
            file.access_condition_id === FILE_ACCESS_CONDITION_OPEN &&
            !!file.date &&
            moment(file.date).isAfter()
        ) {
            securityPolicy = FILE_ACCESS_CONDITION_CLOSED;
        }
    }
    const metadata = {
        dsi_security_inherited: securityInherited ? 1 : 0,
        dsi_security_policy: securityPolicy,
        dsi_label: file.description,
        ...((file.access_condition_id === FILE_ACCESS_CONDITION_OPEN ||
            file.security_policy === FILE_SECURITY_POLICY_PUBLIC) &&
        !moment(file.date).isSame(moment(), 'day')
            ? { dsi_embargo_date: moment(file.date).format(locale.global.embargoDateFormat) }
            : {}),
    };
    return metadata;
};

/**
 * Uploads a file directly into an S3 bucket via API
 *
 * @param {string} pid of object, folder name to where file will be uploaded
 * @param {object} file to be uploaded
 * @param {function} dispatch
 * @param {string} formName the name of the form being used to upload the file
 * @param {array} collections assigned to the record
 *
 * @returns {Promise}
 */
export function putUploadFile(pid, file, dispatch, formName, collections) {
    let retried = false;

    const uploadFile = () =>
        post(FILE_UPLOAD_API(), {
            Key: `${pid}/${file.name}`,
            Metadata: getFileUploadMetadata(file, collections),
        })
            .then(uploadUrl => {
                const extension = file.name
                    .split('.')
                    .pop()
                    .toString()
                    .toLowerCase();
                const headers = {};
                if (MIME_TYPE_WHITELIST.hasOwnProperty(extension)) {
                    headers['Content-Type'] = MIME_TYPE_WHITELIST[extension];
                }
                const options = {
                    headers,
                    onUploadProgress: fileUploadActions.notifyFileUploadProgress(file.name, dispatch),
                    cancelToken: generateCancelToken().token,
                };
                const fileUrl = Array.isArray(uploadUrl) && uploadUrl.length > 0 ? uploadUrl[0] : uploadUrl;
                return put({ apiUrl: fileUrl }, file.fileData, options);
            })
            .then(uploadResponse => {
                fileUploadActions.notifyFileUploadProgress(file.name, dispatch)({ loaded: 1, total: 1 });
                fileUploadActions.markCompletedUpload(formName, file.name)(dispatch);
                return Promise.resolve(uploadResponse);
            })
            .catch(error => {
                if (process.env.ENABLE_LOG) Sentry.captureException(error);
                if (fileUploadActions) {
                    dispatch(fileUploadActions.notifyUploadFailed(file.name));
                }
                if (retried) {
                    return Promise.reject(error);
                } else {
                    retried = true;
                    if (fileUploadActions) {
                        dispatch(fileUploadActions.startFileUpload());
                    }
                    return uploadFile();
                }
            });
    return uploadFile();
}

/**
 * Uploads a list of files
 *
 * @param {string} pid of object, folder name to where file will be uploaded
 * @param {array} files to be uploaded
 * @param {function} dispatch
 * @param {string} formName the name of the form being used to upload the files
 * @param {array} collections assigned to the record
 *
 * @returns {Promise.all}
 */
export function putUploadFiles(pid, files, dispatch, formName = '', collections = []) {
    const filenameList = files && Array.isArray(files) && files.map(item => item.name);
    const checkIfDuplicateExists = w => {
        return new Set(w).size !== w.length;
    };
    const duplicateFileNames = checkIfDuplicateExists(filenameList);
    if (!!duplicateFileNames) {
        Sentry.captureMessage(`Duplicate files found when uploading files for PID ${pid} : ${filenameList}`);
    }
    dispatch(fileUploadActions.startFileUpload());
    const uploadFilesPromises = files.map(file => putUploadFile(pid, file, dispatch, formName, collections));
    return Promise.all(uploadFilesPromises);
}
