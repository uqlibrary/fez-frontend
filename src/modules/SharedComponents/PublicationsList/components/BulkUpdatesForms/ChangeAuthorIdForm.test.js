import React from 'react';
import ChangeAuthorIdForm from './ChangeAuthorIdForm';
import { render, WithRouter, WithReduxStore, fireEvent, waitFor } from 'test-utils';
import * as repositories from 'repositories';

function setup(testProps = {}) {
    const props = {
        recordsSelected: {
            'UQ:123456': {
                rek_pid: 'UQ:123456',
                fez_record_search_key_author: [{ rek_author: 'Test', rek_author_order: 1 }],
                fez_record_search_key_author_id: [{ rek_author_id: null }],
            },
            'UQ:1234': {
                rek_pid: 'UQ:1234',
                fez_record_search_key_author: [
                    { rek_author: 'Test', rek_author_order: 1 },
                    { rek_author: 'Testing', rek_author_order: 2 },
                ],
                fez_record_search_key_author_id: [
                    { rek_author_id: null },
                    { rek_author_id: 2, rek_author_id_order: 2, rek_author_id_id: 22 },
                ],
            },
            'UQ:4321': {
                rek_pid: 'UQ:4321',
                fez_record_search_key_author: [{ rek_author: 'Testing', rek_author_order: 1 }],
                fez_record_search_key_author_id: [
                    { rek_author_id: null },
                    { rek_author_id: 2, rek_author_id_order: 2, rek_author_id_id: 22 },
                ],
            },
        },
        onCancel: jest.fn(),
        ...testProps,
    };

    return render(
        <WithReduxStore>
            <WithRouter>
                <ChangeAuthorIdForm {...props} />
            </WithRouter>
        </WithReduxStore>,
    );
}

describe('ChangeAuthorIdForm', () => {
    beforeEach(() => {
        document.createRange = () => ({
            setStart: () => {},
            setEnd: () => {},
            commonAncestorContainer: {
                nodeName: 'BODY',
                ownerDocument: document,
            },
        });
    });

    it('should correctly search by author name, submit form and display success info', async () => {
        mockApi.onPatch(repositories.routes.NEW_RECORD_API().apiUrl).replyOnce(200, {});
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 111, value: 'Testing', aut_id: 111, aut_org_username: 'uqtest' }],
        });
        const { getByTestId, getByText, queryByTestId } = setup();

        // assert initial state of the form
        expect(getByTestId('search-author-by-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-author-by-helper-text')).toHaveTextContent('This field is required');
        expect(queryByTestId('search-by-rek-author-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toHaveTextContent('This field is required');
        expect(getByTestId('change-author-id-submit')).toHaveAttribute('disabled');

        // interact with the form
        fireEvent.mouseDown(getByTestId('search-author-by-select'));
        await waitFor(() => getByTestId('search-author-by-options'));
        fireEvent.click(getByText('Author name'));

        expect(getByTestId('search-by-rek-author-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-by-rek-author-helper-text')).toHaveTextContent('This field is required');

        fireEvent.change(getByTestId('search-by-rek-author-input'), { target: { value: 'Test' } });

        // assert next state of the form
        expect(queryByTestId('search-by-rek-author-helper-text')).not.toBeInTheDocument();

        expect(getByTestId('alert-warning-change-author-id')).toHaveTextContent(
            '1 of the 3 works you have selected do not match and will not be updated',
        );

        fireEvent.click(getByTestId('rek-author-id-input'));
        fireEvent.change(getByTestId('rek-author-id-input'), { target: { value: 'Testing' } });
        await waitFor(() => getByTestId('rek-author-id-options')).then(() => {
            fireEvent.click(getByText('Testing'));
        });

        // assert next state of the form
        expect(queryByTestId('rek-author-id-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('change-author-id-submit')).not.toHaveAttribute('disabled');

        // submit form
        fireEvent.click(getByTestId('change-author-id-submit'));

        await waitFor(() => expect(getByTestId('alert-done-change-author-id')).toBeInTheDocument());
    });

    it('should correctly search by author name, submit form and display error', async () => {
        mockApi.onPatch(repositories.routes.NEW_RECORD_API().apiUrl).replyOnce(500);
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 111, value: 'Testing', aut_id: 111, aut_org_username: 'uqtest' }],
        });
        const { getByTestId, getByText, queryByTestId } = setup();

        // assert initial state of the form
        expect(getByTestId('search-author-by-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-author-by-helper-text')).toHaveTextContent('This field is required');
        expect(queryByTestId('search-by-rek-author-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toHaveTextContent('This field is required');
        expect(getByTestId('change-author-id-submit')).toHaveAttribute('disabled');

        fireEvent.mouseDown(getByTestId('search-author-by-select'));
        await waitFor(() => getByTestId('search-author-by-options'));
        fireEvent.click(getByText('Author name'));

        expect(getByTestId('search-by-rek-author-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-by-rek-author-helper-text')).toHaveTextContent('This field is required');

        fireEvent.change(getByTestId('search-by-rek-author-input'), { target: { value: 'Test' } });

        // assert next state of the form
        expect(queryByTestId('search-by-rek-author-helper-text')).not.toBeInTheDocument();

        fireEvent.click(getByTestId('rek-author-id-input'));
        fireEvent.change(getByTestId('rek-author-id-input'), { target: { value: 'Testing' } });
        await waitFor(() => getByTestId('rek-author-id-options'));
        fireEvent.click(getByText('Testing'));

        // assert next state of the form
        expect(queryByTestId('rek-author-id-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('change-author-id-submit')).not.toHaveAttribute('disabled');

        // submit form
        fireEvent.click(getByTestId('change-author-id-submit'));

        await waitFor(() => expect(getByTestId('alert-error-change-author-id')).toBeInTheDocument());
    });

    it('should correctly search by author id, submit form and display success info', async () => {
        mockApi.onPatch(repositories.routes.NEW_RECORD_API().apiUrl).replyOnce(200, {});
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 111, value: 'Testing', aut_id: 123, aut_org_username: 'uqtest' }],
        });
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 123, value: 'Testing', aut_id: 111, aut_org_username: 'uqtest' }],
        });
        const { getByTestId, getByText, queryByTestId } = setup({});

        // assert initial state of the form
        expect(getByTestId('search-author-by-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-author-by-helper-text')).toHaveTextContent('This field is required');
        expect(queryByTestId('search-by-rek-author-id-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toHaveTextContent('This field is required');
        expect(getByTestId('change-author-id-submit')).toHaveAttribute('disabled');

        // interact with the form
        fireEvent.mouseDown(getByTestId('search-author-by-select'));
        await waitFor(() => getByTestId('search-author-by-options'));
        fireEvent.click(getByText('Author ID'));

        expect(getByTestId('search-by-rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-by-rek-author-id-helper-text')).toHaveTextContent('This field is required');

        fireEvent.click(getByTestId('search-by-rek-author-id-input'));
        fireEvent.change(getByTestId('search-by-rek-author-id-input'), { target: { value: 'Test' } });
        await waitFor(() => getByTestId('search-by-rek-author-id-options'));
        fireEvent.click(getByText('Testing'));

        // assert next state of the form
        expect(queryByTestId('search-by-rek-author-id-helper-text')).not.toBeInTheDocument();

        fireEvent.click(getByTestId('rek-author-id-input'));
        fireEvent.change(getByTestId('rek-author-id-input'), { target: { value: 'Testing' } });
        await waitFor(() => getByTestId('rek-author-id-options'));
        fireEvent.click(getByText('Testing'));

        // assert next state of the form
        expect(queryByTestId('rek-author-id-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('change-author-id-submit')).not.toHaveAttribute('disabled');

        // submit form
        fireEvent.click(getByTestId('change-author-id-submit'));

        await waitFor(() => expect(getByTestId('alert-done-change-author-id')).toBeInTheDocument());
    });

    it('should correctly search by author id, submit form and display error', async () => {
        mockApi.onPatch(repositories.routes.NEW_RECORD_API().apiUrl).replyOnce(500);
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 111, value: 'Testing', aut_id: 111, aut_org_username: 'uqtest' }],
        });
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 123, value: 'Testing', aut_id: 123, aut_org_username: 'uqtest' }],
        });
        const { getByTestId, getByText, queryByTestId } = setup();

        // assert initial state of the form
        expect(getByTestId('search-author-by-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-author-by-helper-text')).toHaveTextContent('This field is required');
        expect(queryByTestId('search-by-rek-author-id-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('rek-author-id-helper-text')).toHaveTextContent('This field is required');
        expect(getByTestId('change-author-id-submit')).toHaveAttribute('disabled');

        // interact with the form
        fireEvent.mouseDown(getByTestId('search-author-by-select'));
        await waitFor(() => getByTestId('search-author-by-options'));
        fireEvent.click(getByText('Author ID'));

        expect(getByTestId('search-by-rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-by-rek-author-id-helper-text')).toHaveTextContent('This field is required');

        fireEvent.click(getByTestId('search-by-rek-author-id-input'));
        fireEvent.change(getByTestId('search-by-rek-author-id-input'), { target: { value: 'Test' } });
        await waitFor(() => getByTestId('search-by-rek-author-id-options'));
        fireEvent.click(getByText('Testing'));

        // assert next state of the form
        expect(queryByTestId('search-by-rek-author-id-helper-text')).not.toBeInTheDocument();

        fireEvent.click(getByTestId('rek-author-id-input'));
        fireEvent.change(getByTestId('rek-author-id-input'), { target: { value: 'Testing' } });
        await waitFor(() => getByTestId('rek-author-id-options'));
        fireEvent.click(getByText('Testing'));

        // assert next state of the form
        expect(queryByTestId('rek-author-id-helper-text')).not.toBeInTheDocument();
        expect(getByTestId('change-author-id-submit')).not.toHaveAttribute('disabled');

        // submit form
        fireEvent.click(getByTestId('change-author-id-submit'));

        await waitFor(() => expect(getByTestId('alert-error-change-author-id')).toBeInTheDocument());
    });

    it('should correctly clear author ID field', async () => {
        mockApi.onGet(repositories.routes.SEARCH_AUTHOR_LOOKUP_API({ searchQuery: '.*' }).apiUrl).replyOnce(200, {
            data: [{ id: 111, value: 'Testing', aut_id: 123, aut_org_username: 'uqtest' }],
        });
        const { getByTestId, getByText, getAllByTitle } = setup();

        fireEvent.mouseDown(getByTestId('search-author-by-select'));
        await waitFor(() => getByTestId('search-author-by-options'));
        fireEvent.click(getByText('Author ID'));

        expect(getByTestId('search-by-rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-by-rek-author-id-helper-text')).toHaveTextContent('This field is required');

        fireEvent.click(getByTestId('search-by-rek-author-id-input'));
        fireEvent.change(getByTestId('search-by-rek-author-id-input'), { target: { value: 'Test' } });
        await waitFor(() => getByTestId('search-by-rek-author-id-options'));

        fireEvent.click(getByText('Testing'));
        fireEvent.click(getAllByTitle('Clear')[0]);

        expect(getByTestId('search-by-rek-author-id-helper-text')).toBeInTheDocument();
        expect(getByTestId('search-by-rek-author-id-helper-text')).toHaveTextContent('This field is required');
    });
});
