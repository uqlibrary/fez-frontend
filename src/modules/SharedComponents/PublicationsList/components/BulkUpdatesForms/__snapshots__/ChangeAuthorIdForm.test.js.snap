// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeAuthorIdForm should correctly search by author id, submit form and display success info 1`] = `"[{"rek_pid":"UQ:123456"},{"rek_pid":"UQ:1234","fez_record_search_key_author_id":[{"rek_author_id":null},{"rek_author_id":123,"rek_author_id_order":2}]},{"rek_pid":"UQ:4321"}]"`;

exports[`ChangeAuthorIdForm should correctly search by author name, submit form and display success info 1`] = `"[{"rek_pid":"UQ:123456","fez_record_search_key_author_id":[{"rek_author_id":111,"rek_author_id_order":1}]},{"rek_pid":"UQ:1234","fez_record_search_key_author_id":[{"rek_author_id":111,"rek_author_id_order":1},{"rek_author_id":123,"rek_author_id_order":2}]},{"rek_pid":"UQ:4321"}]"`;
